File: lib/src/main/java/net/mm2d/upnp/SsdpServerDelegate.java
/*
 * Copyright(C)  2018 大前良介(OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import net.mm2d.log.Log;
import net.mm2d.util.IoUtils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.InterfaceAddress;
import java.net.MalformedURLException;
import java.net.MulticastSocket;
import java.net.NetworkInterface;
import java.net.SocketTimeoutException;
import java.net.URL;
import java.net.UnknownHostException;
import java.util.List;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * SsdpServerの共通処理を実装するクラス。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
// TODO: SocketChannelを使用した受信(MulticastChannelはAndroid N以降のため保留)
class SsdpServerDelegate implements SsdpServer {
    interface Receiver {
        /**
         * メッセージ受信後の処理、サブクラスにより実装する。
         *
         * @param sourceAddress 送信元アドレス
         * @param data          受信したデータ
         * @param length        受信したデータの長さ
         */
        void onReceive(
                @Nonnull InetAddress sourceAddress,
                @Nonnull byte[] data,
                int length);
    }

    @Nonnull
    private final Receiver mReceiver;
    @Nonnull
    private final NetworkInterface mInterface;
    @Nonnull
    private final InterfaceAddress mInterfaceAddress;

    private final int mBindPort;
    @Nullable
    private MulticastSocket mSocket;
    @Nullable
    private ReceiveTask mReceiveTask;

    /**
     * 使用するインターフェースを指定してインスタンス作成。
     *
     * <p>使用するポートは自動割当となる。
     *
     * @param receiver         パケット受信時にコールされるreceiver
     * @param networkInterface 使用するインターフェース
     */
    SsdpServerDelegate(
            @Nonnull final Receiver receiver,
            @Nonnull final NetworkInterface networkInterface) {
        this(receiver, networkInterface, 0);
    }

    /**
     * 使用するインターフェースとポート指定してインスタンス作成。
     *
     * @param receiver         パケット受信時にコールされるreceiver
     * @param networkInterface 使用するインターフェース
     * @param bindPort         使用するポート
     */
    SsdpServerDelegate(
            @Nonnull final Receiver receiver,
            @Nonnull final NetworkInterface networkInterface,
            final int bindPort) {
        mInterface = networkInterface;
        mInterfaceAddress = findInet4Address(networkInterface.getInterfaceAddresses());
        mBindPort = bindPort;
        mReceiver = receiver;
    }

    // VisibleForTesting
    @Nonnull
    static InterfaceAddress findInet4Address(@Nonnull final List<InterfaceAddress> addressList) {
        for (final InterfaceAddress address : addressList) {
            if (address.getAddress() instanceof Inet4Address) {
                return address;
            }
        }
        throw new IllegalArgumentException("ni does not have IPv4 address.");
    }

    @Override
    @Nonnull
    public InterfaceAddress getInterfaceAddress() {
        return mInterfaceAddress;
    }

    @Override
    public void open() throws IOException {
        if (mSocket != null) {
            close();
        }
        mSocket = createMulticastSocket(mBindPort);
        mSocket.setNetworkInterface(mInterface);
        mSocket.setTimeToLive(4);
    }

    // VisibleForTesting
    @Nonnull
    MulticastSocket createMulticastSocket(final int port) throws IOException {
        return new MulticastSocket(port);
    }

    @Override
    public void close() {
        stop();
        IoUtils.closeQuietly(mSocket);
        mSocket = null;
    }

    @Override
    public void start() {
        if (mReceiveTask != null) {
            stop();
        }
        mReceiveTask = new ReceiveTask(mReceiver, mSocket, mBindPort);
        mReceiveTask.start();
    }

    @Override
    public void stop() {
        stop(false);
    }

    /**
     * 受信スレッドの停止と必要があればJoinを行う。
     *
     * <p>現在の実装ではIO待ちに割り込むことはできないため、
     * joinを指定しても偶然ソケットタイムアウトやソケット受信が発生しないかぎりjoinできない。
     *
     * @param join trueの時スレッドのJoin待ちを行う。
     */
    // VisibleForTesting
    void stop(final boolean join) {
        if (mReceiveTask == null) {
            return;
        }
        mReceiveTask.shutdownRequest(join);
        mReceiveTask = null;
    }

    @Override
    public void send(@Nonnull final SsdpMessage message) {
        if (mSocket == null) {
            return;
        }
        try {
            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
            message.writeData(baos);
            final byte[] data = baos.toByteArray();
            mSocket.send(new DatagramPacket(data, data.length, SSDP_SO_ADDR));
        } catch (final IOException e) {
            Log.w(e);
        }
    }

    /**
     * SsdpMessageのLocationに正常なURLが記述されており、
     * 記述のアドレスとパケットの送信元アドレスに不一致がないか検査する。
     *
     * @param message       確認するSsdpMessage
     * @param sourceAddress 送信元アドレス
     * @return true:送信元との不一致を含めてLocationに不正がある場合。false:それ以外
     */
    public boolean isInvalidLocation(
            @Nonnull final SsdpMessage message,
            @Nonnull final InetAddress sourceAddress) {
        return !isValidLocation(message, sourceAddress);
    }

    private boolean isValidLocation(
            @Nonnull final SsdpMessage message,
            @Nonnull final InetAddress sourceAddress) {
        final String location = message.getLocation();
        if (!Http.isHttpUrl(location)) {
            return false;
        }
        try {
            final InetAddress locationAddress = InetAddress.getByName(new URL(location).getHost());
            return sourceAddress.equals(locationAddress);
        } catch (MalformedURLException | UnknownHostException ignored) {
        }
        return false;
    }

    // VisibleForTesting
    static class ReceiveTask implements Runnable {
        @Nonnull
        private final Receiver mReceiver;
        @Nonnull
        private final MulticastSocket mSocket;
        private final int mBindPort;

        private volatile boolean mShutdownRequest;
        @Nullable
        private Thread mThread;

        /**
         * インスタンス作成
         */
        ReceiveTask(
                @Nonnull final Receiver receiver,
                @Nonnull final MulticastSocket socket,
                final int port) {
            mReceiver = receiver;
            mSocket = socket;
            mBindPort = port;
        }

        /**
         * スレッドを作成して処理を開始する。
         */
        synchronized void start() {
            mShutdownRequest = false;
            mThread = new Thread(this, getClass().getSimpleName());
            mThread.start();
        }

        /**
         * 割り込みを行い、スレッドを終了させる。
         *
         * <p>現在はSocketを使用しているため割り込みは効果がない。
         *
         * @param join Threadのjoin待ちを行う場合はtrue
         */
        synchronized void shutdownRequest(final boolean join) {
            mShutdownRequest = true;
            if (mThread == null) {
                return;
            }
            mThread.interrupt();
            if (join) {
                try {
                    mThread.join(1000);
                } catch (final InterruptedException ignored) {
                }
            }
            mThread = null;
        }

        @Override
        public void run() {
            if (mShutdownRequest) {
                return;
            }
            try {
                joinGroup();
                receiveLoop();
                leaveGroup();
            } catch (final IOException ignored) {
            }
        }

        /**
         * 受信処理を行う。
         *
         * @throws IOException 入出力処理で例外発生
         */
        // VisibleForTesting
        void receiveLoop() throws IOException {
            final byte[] buf = new byte[1500];
            while (!mShutdownRequest) {
                try {
                    final DatagramPacket dp = new DatagramPacket(buf, buf.length);
                    mSocket.receive(dp);
                    if (mShutdownRequest) {
                        break;
                    }
                    mReceiver.onReceive(dp.getAddress(), dp.getData(), dp.getLength());
                } catch (final SocketTimeoutException ignored) {
                }
            }
        }

        /**
         * Joinを行う。
         *
         * <p>特定ポートにBindしていない（マルチキャスト受信ソケットでない）場合は何も行わない
         *
         * @throws IOException Joinコールにより発生
         */
        // VisibleForTesting
        void joinGroup() throws IOException {
            if (mBindPort != 0) {
                mSocket.joinGroup(SSDP_INET_ADDR);
            }
        }

        /**
         * Leaveを行う。
         *
         * <p>特定ポートにBindしていない（マルチキャスト受信ソケットでない）場合は何も行わない
         *
         * @throws IOException Leaveコールにより発生
         */
        // VisibleForTesting
        void leaveGroup() throws IOException {
            if (mBindPort != 0) {
                mSocket.leaveGroup(SSDP_INET_ADDR);
            }
        }
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/common/internal/server/SsdpServerDelegate.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.common.internal.server

import net.mm2d.log.Logger
import net.mm2d.upnp.common.SsdpMessage
import net.mm2d.upnp.common.internal.thread.TaskExecutors
import net.mm2d.upnp.common.internal.thread.ThreadCondition
import net.mm2d.upnp.common.internal.util.closeQuietly
import net.mm2d.upnp.common.util.findInet4Address
import net.mm2d.upnp.common.util.findInet6Address
import net.mm2d.upnp.common.util.toSimpleString
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.*

/**
 * A class that implements the common part of [SsdpServer].
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 *
 * @param taskExecutors taskExecutors
 * @param address Multicast address
 * @param networkInterface network interface
 * @param bindPort port number
 */
internal class SsdpServerDelegate(
    private val taskExecutors: TaskExecutors,
    val address: Address,
    private val networkInterface: NetworkInterface,
    private val bindPort: Int = 0
) : SsdpServer, Runnable {
    val interfaceAddress: InterfaceAddress =
        if (address == Address.IP_V4)
            networkInterface.findInet4Address()
        else
            networkInterface.findInet6Address()
    private var socket: MulticastSocket? = null
    private var receiver: ((sourceAddress: InetAddress, data: ByteArray, length: Int) -> Unit)? = null
    private val threadCondition = ThreadCondition(taskExecutors.server)

    fun setReceiver(receiver: ((sourceAddress: InetAddress, data: ByteArray, length: Int) -> Unit)?) {
        this.receiver = receiver
    }

    fun getSsdpInetAddress(): InetAddress = address.ssdpInetAddress

    fun getSsdpAddressString(): String = address.ssdpAddressString

    fun getLocalAddress(): InetAddress = interfaceAddress.address

    // VisibleForTesting
    @Throws(IOException::class)
    internal fun createMulticastSocket(port: Int): MulticastSocket {
        return MulticastSocket(port).also {
            it.networkInterface = networkInterface
            it.timeToLive = 4
        }
    }

    override fun start() {
        receiver ?: throw IllegalStateException("receiver must be set")
        threadCondition.start(this)
    }

    override fun stop() {
        threadCondition.stop()
        socket.closeQuietly()
    }

    override fun send(messageSupplier: () -> SsdpMessage) {
        taskExecutors.io { sendInner(messageSupplier()) }
    }

    private fun sendInner(message: SsdpMessage) {
        if (!threadCondition.waitReady()) {
            Logger.w("socket is not ready")
            return
        }
        val socket = socket ?: return
        Logger.d { "send from $interfaceAddress:\n$message" }
        try {
            val data = ByteArrayOutputStream().also {
                message.writeData(it)
            }.toByteArray()
            socket.send(DatagramPacket(data, data.size, address.ssdpSocketAddress))
        } catch (e: IOException) {
            Logger.w(e)
        }
    }

    override fun run() {
        val suffix = (if (bindPort == 0) "-ssdp-notify-" else "-ssdp-search-") +
            networkInterface.name + "-" + interfaceAddress.address.toSimpleString()
        Thread.currentThread().let {
            it.name = it.name + suffix
        }
        if (threadCondition.isCanceled()) return
        try {
            val socket = createMulticastSocket(bindPort)
            this.socket = socket
            if (bindPort != 0) {
                socket.joinGroup(getSsdpInetAddress())
            }
            threadCondition.notifyReady()
            receiveLoop(socket)
        } catch (ignored: IOException) {
        } finally {
            if (bindPort != 0) {
                socket?.leaveGroup(getSsdpInetAddress())
            }
            socket.closeQuietly()
            socket = null
        }
    }

    // VisibleForTesting
    @Throws(IOException::class)
    internal fun receiveLoop(socket: MulticastSocket) {
        val buf = ByteArray(1500)
        while (!threadCondition.isCanceled()) {
            try {
                val dp = DatagramPacket(buf, buf.size)
                socket.receive(dp)
                if (threadCondition.isCanceled()) break
                receiver?.invoke(dp.address, dp.data, dp.length)
            } catch (ignored: SocketTimeoutException) {
            }
        }
    }
}
